### DISCUSSION TITLE
Top Reasons NOT to use GitHub Copilot

### DISCUSSION THREAD

**Discussion by colindembovsky** [community/community#58562] [Copilot Conversations]
Generative AI, Copilot, blah blah blah - who needs it? You're the Ultimate Programmer, so why would you want some pretentious "large" language model helping you? You're a lone wolf that don't need nobody (or no _thing_) to "help" you - it will only get in the way of your staggering intellect.

Well this post is just for you - top reasons why do DON'T need GitHub Copilot.

## 1. You prefer writing code in Notepad

GitHub Copilot only supports four IDEs: Visual Studio Code, Visual Studio, IntelliJ and NeoVim. But you prefer to code in Notepad. Or vim. Or emacs. All those plugins and breakpoints and live debugging - it's overrated. You can debug in your head just by looking at the perfect code you wrote. And you can quit vim whenever you want to.

So what if Copilot integrates seamlessly and fades into the background as you code?

## 2. You like writing boilerplate code

Constructors. Getters. Setters. Who needs to think about business problems when you can write _real_ code. I mean, you learned how to do it when you did your Intro to Programming course, so you want to make sure you get your money's worth.

So what if Copilot is _really_ good at writing repetitive, boilerplate code, thereby keeping you focused on solving business problems?

## 3. You know every regex expression.

Only newbies need to test their regex expressions using [regex101](https://regex101.com/). You don't need Copilot's help to validate obscure string formats - you just do it in your head no worries about human error.

So what if Copilot can generate regex and easily dump out obscure formats and formulas so that you don't have to remember or search for them?

## 4. You know every API

Who needs to look up how to invoke common APIs? Once you've seen a Swagger doc you can call any and every method in that API forever.

So what if Copilot knows how to call APIs that millions of developers use daily?

## 5. You like copying and pasting from StackOverflow.

Speaking of searching for stuff - you love StackOverflow! What's better than googling a question and then inevitably landing on StackOverflow where there are a bunch of random answers that may or may not be correct that you can copy from? And who doesn't love renaming all the variables and fixing all the formatting errors (tabs vs spaces anyone)? Not that you need to search for stuff anyway - your infallible memory is a giant library of endless code examples to draw from.

So what if Copilot can get answers for you without you having to leave the IDE... er, file... and follows your naming conventions and styles?

## 6. You don't need unit tests.

Unit tests - that assumes your code could be wrong. And why spend time programming code that tries to break the code you just coded so perfectly? If you _did_ write unit tests, they would be the ultimate tests.

So what if Copilot can quickly generate tests, mocks and find multiple test cases just by analyzing the code you're testing?

## 7. Comments? What for?

You don't have to document your code. Your code is so perfect that people can tell what it's doing just by seeing your code. Besides, no-one else will ever need to look at your code unless it's to learn how to program perfectly. Da Vinci didn't have to "comment" the Mona Lisa, did he?

So what if GitHub Copilot can generate code based on your comments, and that the comments stay to help document your code?

## 8. You'd rather leak IP by pasting code into ChatGPT.

You're not like those Samsung developers that [leaked sensitive information](https://www.techradar.com/news/samsung-workers-leaked-company-secrets-by-using-chatgpt) while copying code into ChatGPT, right? I mean, you'd never be asking an AI for help anyway.

So what if Copilot encrypts data and uses an HTTPS and (at least for Copilot for Business) never keeps any of the data or solutions?

## Conclusion

This GitHub Copilot thing is totally overrated. It's not going to change the way you work or the perfection of the code that you crank out as you consume coffee and cold pizza. No way.

Happy (not) Copiloting!

**Comment by 46102830** [comment]
Es aceptable tu respuesta, pero por desgracia no todos somo unos super genios con esto de los codigos abiertos y que mucho mejor que poder aprender con una ayuda y asi compremder como es que se le puede dar solucion a un problema

**Comment by ng-druid** [comment]
## Harmful To The Environment

Developing on the cloud when a local environment would suffice only adds to carbon emissions. When everyone moves to working on the cloud, using AI demand/resources significantly rise. Unless an org is running on completely green energy every resource added energy added demand thus carbon. Development spirals out of control requiring a single machine to not only multiple machines but network traffic supporting infrastructure to support the demand. Working on CoPilot and/or combing with AI is just more and more energy, demand and emissions.

**Reply by rakotomandimby** [reply]
Having code written quicker reduces workstation uptime or allow more resting time for eyes

**Reply by spenserblack** [reply]
I'd be curious if there could be a study that compares carbon footprint of not using Copilot vs using it, taking into account time saved when developing with an AI assistant. E.g. The carbon footprint of developers of equal skill level creating a simple Twitter clone with and without Copilot.

**Comment by khoadaniel** [comment]
Excellent post, excellent click-bait and excellent twist.

**Reply by naji567** [reply]
G

**Reply by thuyetminh** [reply]
#58608

**Reply by computationalknowledge** [reply]
I thought I was good at sarcasm.... KUDOS to you Sir~!

**Comment by Maxtrade4** [comment]
Wow that's great

**Comment by orl0pl** [comment]
You don't want to pay, because programming is only a hobby (for me)

**Comment by johnwildes** [comment]
Great post @colindembovsky!  When can we expect Terraform and or Bicep support...that would help me out a lot!

**Comment by Viyom29** [comment]
> Generative AI, Copilot, blah blah blah - who needs it? You're the Ultimate Programmer, so why would you want some pretentious "large" language model helping you? You're a lone wolf that don't need nobody (or no _thing_) to "help" you - it will only get in the way of your staggering intellect.
> 
> 
> 
> Well this post is just for you - top reasons why do DON'T need GitHub Copilot.
> 
> 
> 
> ## 1. You prefer writing code in Notepad
> 
> 
> 
> GitHub Copilot only supports four IDEs: Visual Studio Code, Visual Studio, IntelliJ and NeoVim. But you prefer to code in Notepad. Or vim. Or emacs. All those plugins and breakpoints and live debugging - it's overrated. You can debug in your head just by looking at the perfect code you wrote. And you can quit vim whenever you want to.
> 
> 
> 
> So what if Copilot integrates seamlessly and fades into the background as you code?
> 
> 
> 
> ## 2. You like writing boilerplate code
> 
> 
> 
> Constructors. Getters. Setters. Who needs to think about business problems when you can write _real_ code. I mean, you learned how to do it when you did your Intro to Programming course, so you want to make sure you get your money's worth.
> 
> 
> 
> So what if Copilot is _really_ good at writing repetitive, boilerplate code, thereby keeping you focused on solving business problems?
> 
> 
> 
> ## 3. You know every regex expression.
> 
> 
> 
> Only newbies need to test their regex expressions using [regex101](https://regex101.com/). You don't need Copilot's help to validate obscure string formats - you just do it in your head no worries about human error.
> 
> 
> 
> So what if Copilot can generate regex and easily dump out obscure formats and formulas so that you don't have to remember or search for them?
> 
> 
> 
> ## 4. You know every API
> 
> 
> 
> Who needs to look up how to invoke common APIs? Once you've seen a Swagger doc you can call any and every method in that API forever.
> 
> 
> 
> So what if Copilot knows how to call APIs that millions of developers use daily?
> 
> 
> 
> ## 5. You like copying and pasting from StackOverflow.
> 
> 
> 
> Speaking of searching for stuff - you love StackOverflow! What's better than googling a question and then inevitably landing on StackOverflow where there are a bunch of random answers that may or may not be correct that you can copy from? And who doesn't love renaming all the variables and fixing all the formatting errors (tabs vs spaces anyone)? Not that you need to search for stuff anyway - your infallible memory is a giant library of endless code examples to draw from.
> 
> 
> 
> So what if Copilot can get answers for you without you having to leave the IDE... er, file... and follows your naming conventions and styles?
> 
> 
> 
> ## 6. You don't need unit tests.
> 
> 
> 
> Unit tests - that assumes your code could be wrong. And why spend time programming code that tries to break the code you just coded so perfectly? If you _did_ write unit tests, they would be the ultimate tests.
> 
> 
> 
> So what if Copilot can quickly generate tests, mocks and find multiple test cases just by analyzing the code you're testing?
> 
> 
> 
> ## 7. Comments? What for?
> 
> 
> 
> You don't have to document your code. Your code is so perfect that people can tell what it's doing just by seeing your code. Besides, no-one else will ever need to look at your code unless it's to learn how to program perfectly. Da Vinci didn't have to "comment" the Mona Lisa, did he?
> 
> 
> 
> So what if GitHub Copilot can generate code based on your comments, and that the comments stay to help document your code?
> 
> 
> 
> ## 8. You'd rather leak IP by pasting code into ChatGPT.
> 
> 
> 
> You're not like those Samsung developers that [leaked sensitive information](https://www.techradar.com/news/samsung-workers-leaked-company-secrets-by-using-chatgpt) while copying code into ChatGPT, right? I mean, you'd never be asking an AI for help anyway.
> 
> 
> 
> So what if Copilot encrypts data and uses an HTTPS and (at least for Copilot for Business) never keeps any of the data or solutions?
> 
> 
> 
> ## Conclusion
> 
> 
> 
> This GitHub Copilot thing is totally overrated. It's not going to change the way you work or the perfection of the code that you crank out as you consume coffee and cold pizza. No way.
> 
> 
> 
> Happy (not) Copiloting!

**Reply by Viyom29** [reply]
B

https://docs.github.com/en/enterprise-cloud@latest/admin/code-security/managing-github-advanced-security-for-your-enterprise/managing-github-advanced-security-features-for-your-enterprise

**Comment by jsoref** [comment]
9. You don't like risking the wrath of IP lawyers because Copilot has stolen code from a product whose license is incompatible with yours and had you used it your product could be infringing on a landmine of copyrights that you don't understand.

10. You don't like to steal other people's content without properly attributing it.

**Reply by meppu** [reply]
11. Copilot outputs just calls non-existing functions with bunch of nonsense.

12. You are not able to type 3 words and instead waiting for copilot to ~~steal~~ analyse your code and ~~steals someones~~ generate function call for you.

**Comment by chakchak456** [comment]
Now not work?

**Comment by Sbk34** [comment]
I just spent 10 min , that by the way and I'm "that's 10 min of my life I will never get back.". 

      Hello I'm Skielz34, I'm just a real nube to coding and my ego hasn't completely turned me into a rude person because hey I get it just being a level 5 hand finisher in a day in age we're computer 🖥️💻 are doing everything and being a a free hand coder that grew up on old school knowledge, principals, steps,rules , and pure I think genuis well I guess the one like me who stumbled on all things a chat ai server or channel totally glued with awe while trying to figure out a what of all things Flipper Zero. Yeah rite out the box alot of people will be like cool but realistically after it's in there's hands and try to get with it's potential well they'll probably end up tossing it to the side. Sorry I got track any way I m digging this automation and I am truly interested in most of this , but shoot Im 40 , I've been working my ass off physically all my life , but I'm going to have have a real good back up plan cause Im not sure my body is going to hold out , Let's code , learn programing and I myself will use all the tools and time you don't . Ssiss.

**Comment by Sbk34** [comment]
10 reasons why I should.

1. I'm not seeing to many other volunteers.

2. To learn this I am going to need someone or something to teach it .

3. It's going to suggest a lot of stuff I might have thought of.

4. The suggestions are going to be extension of what my brain may not have known was available.

5. I don't have many real friends. Lol
6. Ai is fun to use as a learning tool.
7. This learning tool is I believe progressive in the sense it's always growing the intelligence world.
8. When I wake up late night or early morning and need to get the ole brains wheels working and looking for a explanation why it is is . 
9. The suggestions begin with the user .
10. Im pretty stoked to have found it .

**Comment by sOfTheApes** [comment]
You heard it here first, critical responses are off-topic all the time. There is no such thing as on-topic criticism. It doesn't exist.

**Comment by chrisblanch39** [comment]
#59395 
| Header |
|--------|
| Cell |
| Cell |
| Cell |
| Cell |
| Cell |

**Comment by JudsonOrum** [comment]
Thank you Sir I decided to use GitHub Copilot

**Comment by lalo-mx** [comment]
9. I rather keep using Netbeans

**Comment by colossalchicken** [comment]
GNU Nano is by far the best IDE.

**Comment by jakesant** [comment]
Don't get me wrong I love Copilot and use it every day, but there _are_ valid reasons for not using it. [One study found that code generated by Copilot produced slower than average runtime compared to the canonical solutions](https://arxiv.org/abs/2305.06439). In one study, [40% of the code generated by Copilot had security vulnerabilities listed by CWE.](https://arxiv.org/abs/2108.09293) In all fairness, that study was done when Copilot was still in preview phase.

You can also get too reliant on it. It can make you myopic to other better solutions that there could be as it guides you to a specific path.

+1 for the clickbait but there are valid reasons for not using it, or rather not depending on it too much. I think AI dev tools are the future and we will all be using them more, but just because they generate pretty-decent and valid code quickly quickly doesn't make it the best solution.

**Comment by jakesant** [comment]
> I'd be curious if there could be a study that compares carbon footprint of not using Copilot vs using it, taking into account time saved when developing with an AI assistant. E.g. The carbon footprint of developers of equal skill level creating a simple Twitter clone with and without Copilot.

@spenserblack I found it quite suspicious that your comment and @rakotomandimby's were marked as "off-topic". They are very much on topic.

Re: studies, I actually looked into studies with Copilot last week. I didn't find anything that compared carbon footprint, but I did find some studies comparing the time taken developing with and without Copilot, program runtime, security and so on. Here's a few papers/studies:

* [Measuring the Runtime Performance of Code Produced with GitHub Copilot](https://arxiv.org/abs/2305.06439) - [Daniel Erhabor](https://arxiv.org/search/cs?searchtype=author&query=Erhabor%2C+D), [Sreeharsha Udayashankar](https://arxiv.org/search/cs?searchtype=author&query=Udayashankar%2C+S), [Meiyappan Nagappan](https://arxiv.org/search/cs?searchtype=author&query=Nagappan%2C+M), [Samer Al-Kiswany](https://arxiv.org/search/cs?searchtype=author&query=Al-Kiswany%2C+S)
* [Expectation vs. Experience: Evaluating the Usability of Code Generation Tools Powered by Large Language Models](https://dl.acm.org/doi/abs/10.1145/3491101.3519665) - [Priyan Vaithilingam](https://dl.acm.org/profile/99659260410), [Tianyi Zhang](https://dl.acm.org/profile/99658654086), [Elena L. Glassman](https://dl.acm.org/profile/82658918057)
* [Asleep at the Keyboard? Assessing the Security of GitHub Copilot's Code Contributions](https://arxiv.org/abs/2108.09293) - [Hammond Pearce](https://arxiv.org/search/cs?searchtype=author&query=Pearce%2C+H), [Baleegh Ahmad](https://arxiv.org/search/cs?searchtype=author&query=Ahmad%2C+B), [Benjamin Tan](https://arxiv.org/search/cs?searchtype=author&query=Tan%2C+B), [Brendan Dolan-Gavitt](https://arxiv.org/search/cs?searchtype=author&query=Dolan-Gavitt%2C+B), [Ramesh Karri](https://arxiv.org/search/cs?searchtype=author&query=Karri%2C+R)
* [Assessing the Quality of GitHub Copilot’s Code Generation](https://www.researchgate.net/publication/363267006_Assessing_the_Quality_of_GitHub_Copilot%27s_Code_Generation) - [Burak Yetiştiren](https://www.researchgate.net/profile/Burak-Yetistiren-2), [Eray Tüzün](https://www.researchgate.net/profile/Eray-Tuezuen), [Işık Özsoy](https://www.researchgate.net/profile/Isik_Oezsoy)

The takeaway from those studies and more is that many devs feel Copilot provides a good starting point, and it *does* produce valid (meaning runs without errors) code most of the time.

One of the studies found that ~40% of code it generated had security vulnerabilities listed by Common Weakness Enumeration (CWE). 

Another study couldn't find that it statistically significantly saved time. That study found that developers had to spend time debugging and understanding the code generated too.

And another study examining runtime found that Copilot tended to produce code with a slower than average runtime compared to the canonical solutions to the problems they covered.

**Reply by spenserblack** [reply]
Thanks for finding these studies! In my experience Copilot definitely needs to be guided a bit. The first suggestion isn't always what you want, unless you've written a decent amount of code to heavily hint to it the style of solution you're going for.

>Another study couldn't find that it statistically significantly saved time. That study found that developers had to spend time debugging and understanding the code generated too.

This is probably a good argument for manually writing tests instead of letting Copilot do it. "Copilot tested itself and concluded that it wrote good code." :laughing:
You don't want to have tests that you don't understand or need to debug. And, once you have confidence in your tests, it should be easier to find where Copilot generated unwanted behavior.

In my experience, I've appreciated Copilot the most with highly repetitive behavior. For example:
```go
type Token int

const (
    FOO Token = iota
    BAR
    // a ton of other tokens
)

var tokenNames map[Token]string{
    FOO: "Foo",
    // Start accepting Copilot suggestions here
    BAR: "Bar",
    // a ton of other tokens
}
```
This could have possibly been done with an editor macro, but in this example Copilot does save time IMO. The simplicity leaves little room for Copilot to make a mistake.

So I guess I'm saying that Copilot works best when you understand its limitations and let it speed up simple tasks, saving the complexity for humans.

**Comment by gianpaj** [comment]
# 1. Price

Alternatives are cheaper. Not as fast or good, but $100 per year feels too much.

https://github.com/features/copilot/#pricing

Paying for AI tools subscriptions is getting expensive — OpenAI ChatGPT Plus, Midjourney, etc.

Perhaps implementing a dynamic pricing model similar to YouTube's, based on the user's location, could be a viable solution.

**Comment by Hayderhammami** [comment]
H

**Comment by marniva** [comment]
You got me convinced. I do not need it! 😝

**Comment by octav20** [comment]
11- Tabnine exist

**Comment by StevenCanDrink** [comment]
- it making the Sloth inside us activated, but it is helpful, using it like assistants, dont abuse them to be slave for AI in general and for Github Pilot in Specific
